WcaOnRails/app/javascript/edit-schedule/SchedulesEditor/calendar-utils.js

```diff
@@ -0,0 +1,230 @@
+import { rootRender } from 'edit-schedule'
+import { newActivityId, defaultDurationFromActivityCode } from '../utils'
+import {
+  activityIndexInArray,
+  roomWcifFromId,
+  venueWcifFromRoomId,
+} from 'wca/wcif-utils'
+import { scheduleElementSelector } from './fullcalendar'
+
+const tzConverterHandlers = {
+};
+
+const calendarHandlers = {
+};
+
+function handleAddActivityToCalendar(reactElem, activityData, renderItOnCalendar) {
+  let currentEventSelected = selectedEventInCalendar();
+  let roomSelected = roomWcifFromId(reactElem.props.scheduleWcif, reactElem.state.selectedRoom);
+  if (roomSelected) {
+    let newActivity = {
+      id: activityData.id || newActivityId(),
+      name: activityData.name,
+      activityCode: activityData.activityCode,
+      childActivities: [],
+    };
+    if (activityData.startTime && activityData.endTime) {
+      newActivity.startTime = activityData.startTime;
+      newActivity.endTime = activityData.endTime;
+    } else if (currentEventSelected) {
+      let newStart = currentEventSelected.end.clone();
+      newActivity.startTime = momentToIso(newStart);
+      let newEnd = newStart.add(defaultDurationFromActivityCode(newActivity.activityCode), "m");
+      newActivity.endTime = momentToIso(newEnd);
+    } else {
+      // Do nothing, user cliked an event without any event selected.
+      return;
+    }
+    roomSelected.activities.push(newActivity);
+    if (renderItOnCalendar) {
+      // Cloning the object here: activityToFcEvent modifies in place,
+      // and we don't want event.selected to propagate in the WCIF!
+      let fcEvent = Object.assign({}, newActivity);
+      fcEvent = activityToFcEvent(fcEvent);
+      singleSelectEvent(fcEvent);
+      $(scheduleElementSelector).fullCalendar("renderEvent", fcEvent);
+    }
+    // update list of activityCode used, and rootRender to display the save message
+    reactElem.setState({ usedActivityCodeList: [...reactElem.state.usedActivityCodeList, newActivity.activityCode] }, rootRender());
+  }
+}
+
+function handleEventModifiedInCalendar(reactElem, event) {
+  let room = roomWcifFromId(reactElem.props.scheduleWcif, reactElem.state.selectedRoom);
+  let activityIndex = activityIndexInArray(room.activities, event.id);
+  if (activityIndex < 0) {
+    alert("This is very very BAD, I couldn't find an activity matching the modified event!");
```

Comment:
I think it might be a better pattern to instead do a `throw new Error()` here. The way this is written we'll proceed onto line 58, and I don't want to think about what happens after that!

If you're concerned about the users learning that something went wrong, then I think we should set up one of React's error boundaries (https://hackernoon.com/error-boundaries-in-react-16-32fb8e185a3). Please don't worry about adding an error boundary to this PR though, it's definitely big enough already!